### ISSUE TITLE
NVDA speech cuts off at end of lines

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#2140]
**Reported by challsworth2 on 2012-03-02 17:25**
Using Espeak, NVDA speech cuts off at the end of lines. This does not happen with Microsoft Speech API version 5.

**Comment by jcsteh** [comment]
There are also cases where the start of utterances are truncated; e.g. when using certain headsets where nothing else has the audio device open.

Things to investigate:

* Is keeping the audio device open the entire time really bad?
 * Does it negatively impact power consumption?
 * Does it prevent standby? #400 suggests it does, but that should be re-tested. Also, that was probably XP. Does this affect later Windows?
 * I have a headset which seems to refuse to play any audio at all if you plug it in while something is holding audio open. Is this actually valid?
* Are there any recommendations about such things? What do other applications which must play non-continuous chunks of audio do?
* We could close the audio device only after a certain time of inactivity.
 * Is this enough to prevent truncation in all cases or do we also need to inject trailing silence? Keeping the device open seems to be enough on devices I have here, but I believe this isn't true for Windows 10 with Realtek audio. @tspivey might have more info.
* If we do close the device after inactivity, this still means the start of utterances might be truncated after the timeout.
 * Is this acceptable?
 * Is there any way to work around this other than prepending silence?
 * I guess we could just prepend silence only when we have to open the device.
* When the device is open, do the start of utterances ever get truncated?
 * If so, we'll need to prepend silence in all cases... but that would cause latency.

CC @Camlorn, @MichaelDCurran, @feerrenrut. In looking at this, we should also consider tones (#400).

**Comment by camlorn** [comment]
The easiest fix for this is probably going to be to just append 100MS of zeros to the end of the audio.  The following Python list comprehension produces the right amount of data:

```
[0.0 for i in xrange(channels*sr*duration)]
```

Where `duration` is in seconds.  These arrays can be preallocated and cached pretty trivially, so performance isn't really an issue.

I think this is doable, but it's been a while since I looked at it.  This can probably just be done by the method whose name I believe is `idle`, as other cases will cut it off anyway.

If audio is being cut off at the beginning, this is just something else to add to the "do we need an audio settings dialog?" discussion.  There is no other way to fix this that I am aware of besides adding latency.

There is a process called convolution which most audio people will be familiar with.  If you want to make a latency free convolution effect, you can do it by dropping the first samples and dealing with some artifacts on device open.  I personally think this is a horrible idea, but I could see some enterprising programmer somewhere going "I know!" and doing it.  So even though I haven't seen it, I could believe in it happening for someone.

Holding the device open may be sufficient for Vista and later, but may not work for XP.  Appending the zeros should, and has the added advantage of building in the waiting without much further work.  This is something that can probably also be improved with #5096, so I'm mentioning that ticket here.

**Comment by Adriani90** [comment]
@challsworth2 is this still an issue?

@michaelDCurran your thoughts are very apprecaited, thanks.

**Comment by Adriani90** [comment]
@camlorn any updates on this?
@leonardder I know you have created an issue which is somehow related, I had this issue as well. See #8017. @mltony created a wonderful add-on for solving this. In my case it works very reliable.

**Comment by Adriani90** [comment]
Closing in favor of #10185. And this might be fixed by #11024.
